File: PlikShare.IntegrationTests/Infrastructure/Apis/GeneralSettingsApi.cs
using Flurl.Http;
using PlikShare.GeneralSettings;
using PlikShare.GeneralSettings.Contracts;
using PlikShare.GeneralSettings.SignUpCheckboxes.CreateOrUpdate.Contracts;
using PlikShare.Users.PermissionsAndRoles;

namespace PlikShare.IntegrationTests.Infrastructure.Apis;

public class GeneralSettingsApi(IFlurlClient flurlClient, string appUrl)
{
    public async Task<GetApplicationSettingsResponse> Get(SessionAuthCookie? cookie)
    {
        return await flurlClient.ExecuteGet<GetApplicationSettingsResponse>(
            appUrl: appUrl,
            apiPath: "api/general-settings",
            cookie: cookie);
    }

    public async Task SetApplicationSignUp(
        AppSettings.SignUpSetting value,
        SessionAuthCookie? cookie,
        AntiforgeryCookies antiforgery)
    {
        await flurlClient.ExecutePatch(
            appUrl: appUrl,
            apiPath: "api/general-settings/application-sign-up",
            request: new SetSettingRequest
            {
                Value = value.Value
            },
            cookie: cookie,
            antiforgery: antiforgery);
    }

    public async Task SetApplicationName(
        string? name,
        SessionAuthCookie? cookie,
        AntiforgeryCookies antiforgery)
    {
        await flurlClient.ExecutePatch(
            appUrl: appUrl,
            apiPath: "api/general-settings/application-name",
            request: new SetSettingRequest
            {
                Value = name
            },
            cookie: cookie,
            antiforgery: antiforgery);
    }

    public async Task SetNewUserDefaultMaxWorkspaceNumber(
        int? value,
        SessionAuthCookie? cookie,
        AntiforgeryCookies antiforgery)
    {
        await flurlClient.ExecutePatch(
            appUrl: appUrl,
            apiPath: "api/general-settings/new-user-default-max-workspace-number",
            request: new SetNewUserDefaultMaxWorkspaceNumberRequestDto
            {
                Value = value
            },
            cookie: cookie,
            antiforgery: antiforgery);
    }

    public async Task SetNewUserDefaultMaxWorkspaceSizeInBytes(
        long? value,
        SessionAuthCookie? cookie,
        AntiforgeryCookies antiforgery)
    {
        await flurlClient.ExecutePatch(
            appUrl: appUrl,
            apiPath: "api/general-settings/new-user-default-max-workspace-size-in-bytes",
            request: new SetNewUserDefaultMaxWorkspaceSizeInBytesRequestDto
            {
                Value = value
            },
            cookie: cookie,
            antiforgery: antiforgery);
    }

    public async Task SetNewUserDefaultMaxWorkspaceTeamMembers(
        int? value,
        SessionAuthCookie? cookie,
        AntiforgeryCookies antiforgery)
    {
        await flurlClient.ExecutePatch(
            appUrl: appUrl,
            apiPath: "api/general-settings/new-user-default-max-workspace-team-members",
            request: new SetNewUserDefaultMaxWorkspaceTeamMembersRequestDto
            {
                Value = value
            },
            cookie: cookie,
            antiforgery: antiforgery);
    }

    public async Task SetNewUserDefaultPermissionsAndRoles(
        UserPermissionsAndRolesDto request,
        SessionAuthCookie? cookie,
        AntiforgeryCookies antiforgery)
    {
        await flurlClient.ExecutePatch(
            appUrl: appUrl,
            apiPath: "api/general-settings/new-user-default-permissions-and-roles",
            request: request,
            cookie: cookie,
            antiforgery: antiforgery);
    }

    public async Task SetAlertOnNewUserRegistered(
        bool isTurnedOn,
        SessionAuthCookie? cookie,
        AntiforgeryCookies antiforgery)
    {
        await flurlClient.ExecutePatch(
            appUrl: appUrl,
            apiPath: "api/general-settings/alert-on-new-user-registered",
            request: new SetAlertSettingReuqest
            {
                IsTurnedOn = isTurnedOn
            },
            cookie: cookie,
            antiforgery: antiforgery);
    }

    public async Task<CreateOrUpdateSignUpCheckboxResponseDto> CreateOrUpdateSignUpCheckbox(
        CreateOrUpdateSignUpCheckboxRequestDto request,
        SessionAuthCookie? cookie,
        AntiforgeryCookies antiforgery)
    {
        return await flurlClient.ExecutePost<CreateOrUpdateSignUpCheckboxResponseDto, CreateOrUpdateSignUpCheckboxRequestDto>(
            appUrl: appUrl,
            apiPath: "api/general-settings/sign-up-checkboxes",
            request: request,
            cookie: cookie,
            antiforgery: antiforgery);
    }

    public async Task DeleteSignUpCheckbox(
        int signUpCheckboxId,
        SessionAuthCookie? cookie,
        AntiforgeryCookies antiforgery)
    {
        await flurlClient.ExecuteDelete(
            appUrl: appUrl,
            apiPath: $"api/general-settings/sign-up-checkboxes/{signUpCheckboxId}",
            cookie: cookie,
            antiforgery: antiforgery);
    }
}

File: Frontend/src/app/services/general-settings.api.ts
import { HttpClient, HttpEvent, HttpHeaders } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { firstValueFrom, Observable } from "rxjs";
import { UserPermissionsAndRolesDto } from "./users.api";

export type ApplicationSingUp = 'everyone' | 'only-invited-users';

export interface GetApplicationSettingsResponse {
    applicationSignUp: ApplicationSingUp;
    termsOfService: string | null;
    privacyPolicy: string | null;
    applicationName: string | null;
    signUpCheckboxes: SignUpCheckboxDto[];
    newUserDefaultMaxWorkspaceNumber: number | null;
    newUserDefaultMaxWorkspaceSizeInBytes: number | null;
    newUserDefaultPermissionsAndRoles: {
        isAdmin: boolean;
        canAddWorkspace: boolean;
        canManageGeneralSettings: boolean;
        canManageUsers: boolean;
        canManageStorages: boolean;
        canManageEmailProviders: boolean;
    },
    alertOnNewUserRegistered: boolean;
}

export interface SignUpCheckboxDto {
    id: number;
    text: string;
    isRequired: boolean;
}

export interface SetApplicationSignUpRequest {
    value: ApplicationSingUp;
}

export interface SetApplicationNameRequest {
    value: string | null;
}

export interface CreateOrUpdateSignUpCheckboxRequest {
    id: number | null;
    text: string;
    isRequired: boolean;
}

export interface CreateOrUpdateSignUpCheckboxResponse {
    newId: number;
}

export interface SetNewUserDefaultMaxWorkspaceNumberRequestDto {
    value: number | null;
}

export interface SetNewUserDefaultMaxWorkspaceSizeInBytesRequestDto {
    value: number | null;
}

export interface SetAlertSettingRequest {
    isTurnedOn: boolean;
}

@Injectable({
    providedIn: 'root'
})
export class GeneralSettingsApi {
    constructor(
        private _http: HttpClient) {        
    }

    public async deleteSignUpCheckobx(id: number) {
        const response = this
            ._http
            .delete(`/api/general-settings/sign-up-checkboxes/${id}`);

        await firstValueFrom(response);
    }

    public async createOrUpdateSignUpCheckbox(request: CreateOrUpdateSignUpCheckboxRequest): Promise<CreateOrUpdateSignUpCheckboxResponse> {
        const response = this
            ._http
            .post<CreateOrUpdateSignUpCheckboxResponse>(
                `/api/general-settings/sign-up-checkboxes`, request);

        return await firstValueFrom(response);
    }

    public async getAppSettings(): Promise<GetApplicationSettingsResponse> {
        const call = this
            ._http
            .get<GetApplicationSettingsResponse>(
                `/api/general-settings`, {
                headers: new HttpHeaders({
                    'Content-Type':  'application/json'
                })
            });

        return await firstValueFrom(call);
    }

    public async setApplicationSingUp(request: SetApplicationSignUpRequest) {
        const call = this
            ._http
            .patch(
                `/api/general-settings/application-sign-up`,request, {
                headers: new HttpHeaders({
                    'Content-Type':  'application/json'
                })
            });

        await firstValueFrom(call);
    }

    public async setApplicationName(request: SetApplicationNameRequest) {
        const call = this
            ._http
            .patch(
                `/api/general-settings/application-name`,request, {
                headers: new HttpHeaders({
                    'Content-Type':  'application/json'
                })
            });

        await firstValueFrom(call);
    }

    public uploadTermsOfService(file: File): Observable<HttpEvent<Object>> {
        const formData = new FormData();

        formData.append('file', file, file.name);

        return this
            ._http
            .post(
                `/api/general-settings/terms-of-service`,formData, {
                reportProgress: true,
                observe: "events"
            });
    }

    public async deleteTermsOfService(): Promise<void> {
        const call = this
            ._http
            .delete(`/api/general-settings/terms-of-service`);

        await firstValueFrom(call);
    }

    public uploadPrivacyPolicy(file: File): Observable<HttpEvent<Object>> {
        const formData = new FormData();

        formData.append('file', file, file.name);

        return this
            ._http
            .post(
                `/api/general-settings/privacy-policy`,formData, {
                reportProgress: true,
                observe: "events"
            });
    }

    public async deletePrivacyPolicy(): Promise<void> {
        const call = this
            ._http
            .delete(`/api/general-settings/privacy-policy`);

        await firstValueFrom(call);
    }

    public async setNewUserDefaultMaxWorkspaceNumber(request: SetNewUserDefaultMaxWorkspaceNumberRequestDto) {
        const call = this
            ._http
            .patch(
                `/api/general-settings/new-user-default-max-workspace-number`, request, {
                headers: new HttpHeaders({
                    'Content-Type': 'application/json'
                })
            });
    
        await firstValueFrom(call);
    }
    
    public async setNewUserDefaultMaxWorkspaceSizeInBytes(request: SetNewUserDefaultMaxWorkspaceSizeInBytesRequestDto) {
        const call = this
            ._http
            .patch(
                `/api/general-settings/new-user-default-max-workspace-size-in-bytes`, request, {
                headers: new HttpHeaders({
                    'Content-Type': 'application/json'
                })
            });
    
        await firstValueFrom(call);
    }
    
    public async setNewUserDefaultPermissionsAndRoles(request: UserPermissionsAndRolesDto) {
        const call = this
            ._http
            .patch(
                `/api/general-settings/new-user-default-permissions-and-roles`, request, {
                headers: new HttpHeaders({
                    'Content-Type': 'application/json'
                })
            });
    
        await firstValueFrom(call);
    }

    public async setAlertOnNewUserRegistered(request: SetAlertSettingRequest) {
        const call = this
            ._http
            .patch(
                `/api/general-settings/alert-on-new-user-registered`, request, {
                headers: new HttpHeaders({
                    'Content-Type': 'application/json'
                })
            });
    
        await firstValueFrom(call);
    }
}